### ISSUE TITLE
Rewarded Video Memory Leak

### ISSUE THREAD

**Issue by zeropol2** [mopub/mopub-android-sdk#446]
One line summary of the issue here

Memory Leak is occured.

#### Steps to reproduce the behavior:

CASE 1. Finish activity by back button after `MoPubRewardedVideos.loadRewardedVideo` called
```
┬───
│ GC Root: Input or output parameters in native code
│
├─ android.os.MessageQueue instance
│    Leaking: NO (MessageQueue#mQuitting is false)
│    HandlerThread: "main"
│    ↓ MessageQueue.mMessages
│                   ~~~~~~~~~
├─ android.os.Message instance
│    Leaking: UNKNOWN
│    Retaining 131672 bytes in 1314 objects
│    Message.what = 0
│    Message.when = 163469307 (108 ms after heap dump)
│    Message.obj = null
│    Message.callback = instance @324600264 of com.iab.omid.library.mopub.
│    walking.TreeWalker$3
│    ↓ Message.next
│              ~~~~
├─ android.os.Message instance
│    Leaking: UNKNOWN
│    Retaining 131612 bytes in 1313 objects
│    Message.what = 0
│    Message.when = 177761842 (14292643 ms after heap dump)
│    Message.obj = null
│    Message.callback = instance @323565824 of com.mopub.mobileads.
│    -$$Lambda$MoPubFullscreen$V0L-nqITwRU17B2IJLBPDlzktRk
│    ↓ Message.callback
│              ~~~~~~~~
├─ com.mopub.mobileads.-$$Lambda$MoPubFullscreen$V0L-nqITwRU17B2IJLBPDlzktRk
│  instance
│    Leaking: UNKNOWN
│    Retaining 131527 bytes in 1311 objects
│    ↓ -$$Lambda$MoPubFullscreen$V0L-nqITwRU17B2IJLBPDlzktRk.f$0
│                                                            ~~~
├─ com.mopub.mobileads.MoPubFullscreen instance
│    Leaking: UNKNOWN
│    Retaining 131515 bytes in 1310 objects
│    mContext instance of com.example.mopubsample.MainActivity with mDestroyed
│    = true
│    ↓ MoPubFullscreen.mContext
│                      ~~~~~~~~
╰→ com.example.mopubsample.MainActivity instance
     Leaking: YES (ObjectWatcher was watching this because com.example.
     mopubsample.MainActivity received Activity#onDestroy() callback and
     Activity#mDestroyed is true)
     Retaining 125064 bytes in 1238 objects
     key = d29c04d2-4204-44bf-8e26-cbf794757498
     watchDurationMillis = 10984
     retainedDurationMillis = 5981
     mApplication instance of android.app.Application
     mBase instance of androidx.appcompat.view.ContextThemeWrapper, not
     wrapping known Android context

METADATA

Build.VERSION.SDK_INT: 27
Build.MANUFACTURER: Huawei
LeakCanary version: 2.5
App process name: com.example.mopubsample
Stats: LruCache[maxSize=3000,hits=1482,misses=38333,hitRate=3%]
RandomAccess[bytes=2036712,reads=38333,travel=15329758263,range=13017509,size=18
160663]
Analysis duration: 11209 ms 
```

CASE 2. Close the rewarded video.
```
┬───
│ GC Root: Input or output parameters in native code
│
├─ android.os.MessageQueue instance
│    Leaking: NO (MessageQueue#mQuitting is false)
│    HandlerThread: "main"
│    ↓ MessageQueue.mMessages
│                   ~~~~~~~~~
├─ android.os.Message instance
│    Leaking: UNKNOWN
│    Retaining 131672 bytes in 1314 objects
│    Message.what = 0
│    Message.when = 163469307 (108 ms after heap dump)
│    Message.obj = null
│    Message.callback = instance @324600264 of com.iab.omid.library.mopub.
│    walking.TreeWalker$3
│    ↓ Message.next
│              ~~~~
├─ android.os.Message instance
│    Leaking: UNKNOWN
│    Retaining 131612 bytes in 1313 objects
│    Message.what = 0
│    Message.when = 177761842 (14292643 ms after heap dump)
│    Message.obj = null
│    Message.callback = instance @323565824 of com.mopub.mobileads.
│    -$$Lambda$MoPubFullscreen$V0L-nqITwRU17B2IJLBPDlzktRk
│    ↓ Message.callback
│              ~~~~~~~~
├─ com.mopub.mobileads.-$$Lambda$MoPubFullscreen$V0L-nqITwRU17B2IJLBPDlzktRk
│  instance
│    Leaking: UNKNOWN
│    Retaining 131527 bytes in 1311 objects
│    ↓ -$$Lambda$MoPubFullscreen$V0L-nqITwRU17B2IJLBPDlzktRk.f$0
│                                                            ~~~
├─ com.mopub.mobileads.MoPubFullscreen instance
│    Leaking: UNKNOWN
│    Retaining 131515 bytes in 1310 objects
│    mContext instance of com.example.mopubsample.MainActivity with mDestroyed
│    = true
│    ↓ MoPubFullscreen.mContext
│                      ~~~~~~~~
╰→ com.example.mopubsample.MainActivity instance
     Leaking: YES (ObjectWatcher was watching this because com.example.
     mopubsample.MainActivity received Activity#onDestroy() callback and
     Activity#mDestroyed is true)
     Retaining 125064 bytes in 1238 objects
     key = d29c04d2-4204-44bf-8e26-cbf794757498
     watchDurationMillis = 10984
     retainedDurationMillis = 5981
     mApplication instance of android.app.Application
     mBase instance of androidx.appcompat.view.ContextThemeWrapper, not
     wrapping known Android context

METADATA

Build.VERSION.SDK_INT: 27
Build.MANUFACTURER: Huawei
LeakCanary version: 2.5
App process name: com.example.mopubsample
Stats: LruCache[maxSize=3000,hits=1482,misses=38333,hitRate=3%]
RandomAccess[bytes=2036712,reads=38333,travel=15329758263,range=13017509,size=18
160663]
Analysis duration: 11209 ms 
```

My source code is

```kotlin
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        MoPub.onCreate(this)
        setContentView(R.layout.activity_main)
        initMoPubSdk()
    }

    override fun onStart() {
        super.onStart()
        MoPub.onStart(this)
    }

    override fun onResume() {
        super.onResume()
        MoPub.onResume(this)
    }

    override fun onPause() {
        super.onPause()
        MoPub.onPause(this)
    }

    override fun onStop() {
        super.onStop()
        MoPub.onStop(this)
    }

    override fun onDestroy() {
        MoPubRewardedVideos.setRewardedVideoListener(null)
        super.onDestroy()
        MoPub.onDestroy(this)
    }


    private fun initMoPubSdk() {
        val configBuilder: SdkConfiguration.Builder = SdkConfiguration.Builder(TEST_AD_UNIT_ID_REWARDED_VIDEO)
        configBuilder.withLogLevel(MoPubLog.LogLevel.DEBUG)
        MoPub.initializeSdk(this, configBuilder.build()) {
            MoPubRewardedVideos.setRewardedVideoListener(object : MoPubRewardedVideoListener {
                override fun onRewardedVideoLoadSuccess(adUnitId: String) {
                    MoPubRewardedVideos.showRewardedVideo(TEST_AD_UNIT_ID_REWARDED_VIDEO)
                }

                override fun onRewardedVideoLoadFailure(adUnitId: String, errorCode: MoPubErrorCode) {
                }

                override fun onRewardedVideoStarted(adUnitId: String) {
                }

                override fun onRewardedVideoPlaybackError(adUnitId: String, errorCode: MoPubErrorCode) {
                }

                override fun onRewardedVideoClicked(adUnitId: String) {
                }

                override fun onRewardedVideoClosed(adUnitId: String) {
                }

                override fun onRewardedVideoCompleted(adUnitIds: MutableSet<String>, reward: MoPubReward) {
                }

            })
            MoPubRewardedVideos.loadRewardedVideo(TEST_AD_UNIT_ID_REWARDED_VIDEO)
        }
    }

    companion object {
        private const val TEST_AD_UNIT_ID_REWARDED_VIDEO = "920b6145fb1546cf8b5cf2ac34638bb7"
    }
}
```

[MoPubSample.zip](https://github.com/mopub/mopub-android-sdk/files/5647286/MoPubSample.zip)

**Comment by hmertkahraman** [comment]
Hi @zeropol2 ,

Thanks for reaching out and providing the details. 

Could you please remove the lifecycle statements of MoPub which are: `MoPub.onStart(this)` , `MoPub.onResume(this)` , ... and retry?

You can refer to our sample application on our repository located [here](https://github.com/mopub/mopub-android-sdk/tree/master/mopub-sample), (specifically to `MoPubSampleActivity.java`) for the best practice implementations.

Thanks,
Mert

**Comment by zeropol2** [comment]
@hmertkahraman 
I deleted the lifecycle callback.

```kotlin
package com.example.mopubsample

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import com.mopub.common.MoPub
import com.mopub.common.MoPubReward
import com.mopub.common.SdkConfiguration
import com.mopub.common.logging.MoPubLog
import com.mopub.mobileads.MoPubErrorCode
import com.mopub.mobileads.MoPubRewardedVideoListener
import com.mopub.mobileads.MoPubRewardedVideos

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initMoPubSdk()
    }


    override fun onDestroy() {
        MoPubRewardedVideos.setRewardedVideoListener(null)
        super.onDestroy()
    }


    private fun initMoPubSdk() {
        val configBuilder: SdkConfiguration.Builder = SdkConfiguration.Builder(TEST_AD_UNIT_ID_REWARDED_VIDEO)
        configBuilder.withLogLevel(MoPubLog.LogLevel.DEBUG)
        MoPub.initializeSdk(this, configBuilder.build()) {
            MoPubRewardedVideos.setRewardedVideoListener(object : MoPubRewardedVideoListener {
                override fun onRewardedVideoLoadSuccess(adUnitId: String) {
                    MoPubRewardedVideos.showRewardedVideo(TEST_AD_UNIT_ID_REWARDED_VIDEO)
                }

                override fun onRewardedVideoLoadFailure(adUnitId: String, errorCode: MoPubErrorCode) {
                }

                override fun onRewardedVideoStarted(adUnitId: String) {
                }

                override fun onRewardedVideoPlaybackError(adUnitId: String, errorCode: MoPubErrorCode) {
                }

                override fun onRewardedVideoClicked(adUnitId: String) {
                }

                override fun onRewardedVideoClosed(adUnitId: String) {
                }

                override fun onRewardedVideoCompleted(adUnitIds: MutableSet<String>, reward: MoPubReward) {
                }

            })
            MoPubRewardedVideos.loadRewardedVideo(TEST_AD_UNIT_ID_REWARDED_VIDEO)
        }
    }

    companion object {
        private const val TEST_AD_UNIT_ID_REWARDED_VIDEO = "920b6145fb1546cf8b5cf2ac34638bb7"
    }
}
```

But the result is same.

Also the docs said that I need to implement the lifecylce callbacks.

https://developers.mopub.com/publishers/android/rewarded-video/#step-5-optionally-implement-lifecycle-callbacks

**Comment by hmertkahraman** [comment]
Hi @zeropol2 ,

Thank you very much for testing this out, we wanted to see if any changes occurred on your environment. We will investigate this issue further and get back to you.

Mert

**Comment by hmertkahraman** [comment]
Hi @zeropol2 ,

This leak problem has been addressed in MoPub 5.16.0, please see the Changelog [here](https://github.com/mopub/mopub-android-sdk/releases/tag/v5.16.0). Please let us know if you still see similar problems or have any other questions.

Thanks,
Mert
